radar-upload-backend/src/main/java/org/radarbase/upload/doa/RecordRepositoryImpl.kt

```diff
@@ -209,38 +232,64 @@ class RecordRepositoryImpl(@Context private var em: javax.inject.Provider<Entity
-                    "Found ${existingMetadata.status}, expected ${RecordStatus.QUEUED}")
-        }
-
-        if ((metadata.status == RecordStatus.SUCCEEDED.toString()
-                || metadata.status == RecordStatus.FAILED.toString())
-                && existingMetadata.status != RecordStatus.PROCESSING) {
-            throw ConflictException("incompatible_status", "Record cannot be updated: Conflict in record meta-data status. " +
-                    "Found ${existingMetadata.status}, expected ${RecordStatus.PROCESSING}")
-        }
         logger.debug("Updating record $id status from ${existingMetadata.status} to ${metadata.status}")
         existingMetadata.apply {
-            status = RecordStatus.valueOf(metadata.status)
+            metadata.status?.let { newStatus ->
+                if (!allowedStateTransition(existingMetadata.status, newStatus)) {
+                    throw ConflictException("incompatible_status", "Record cannot be updated: Conflict in record meta-data status. " +
+                            "Cannot transition from state ${existingMetadata.status} to ${metadata.status}")
+                }
+
+                status = RecordStatus.valueOf(newStatus)
+            }
             message = metadata.message
         }.update()
 
         merge<RecordMetadata>(existingMetadata)
     }
 
-    override fun delete(record: Record) = em.get().transact { remove(record) }
+    override fun delete(record: Record, revision: Int) = em.get().transact {
+        refresh(record)
+        if (record.metadata.revision != revision) {
+            throw ConflictException("incompatible_revision", "Revision in metadata ${record.metadata.revision} does not match provided revision $revision")
+        }
+        if (record.metadata.status == RecordStatus.PROCESSING) {
+            throw ConflictException("incompatible_status", "Cannot delete record ${record.id}: it is currently being processed.")
+        }
+        remove(record)
+    }
 
     override fun readMetadata(id: Long): RecordMetadata? = em.get().transact { find(RecordMetadata::class.java, id) }
 
     companion object {
         private val logger = LoggerFactory.getLogger(RecordRepositoryImpl::class.java)
+
+        private fun allowedStateTransition(from: RecordStatus, to: String): Boolean {
+            val toStatus = try {
+                RecordStatus.valueOf(to)
+            } catch (ex: IllegalArgumentException) {
+                throw BadRequestException("unknown_status", "Record status $from is not a known status. Use one of: ${RecordStatus.values().joinToString()}.")
+            }
+
+            return toStatus == from
+                    || toStatus in allowedStateTransitions[from]
+                    ?: throw BadRequestException("unknown_status", "Record status $from is not a known status. Use one of: ${RecordStatus.values().joinToString()}.")
```

Comment:
I guess you forgot to change the text here. `?: throw BadRequestException("incompatible_update", "Record status $from cannot be update to $to")`

Reply:
This is another error: the `from` status is not listed in the `allowedStateTransitions` map. It should not occur at the moment, it's just as a fail-safe.